[SPIKE: CORL-2560] Create a story tree lookup for computing next unseen comment

nick-funk opened the pull request:
## What does this PR do?

- Creates a story tree of the entire comment hierarchy on the story document.
- Adds a new `nextUnseenComment` query.

## What changes to the GraphQL/Database Schema does this PR introduce?

- Adds a `generateStoryTree` mutation
- Adds a `nextUnseenComment` query

## Does this PR introduce any new environment variables or feature flags? 

No

## If any indexes were added, were they added to `INDEXES.md`?

No new indexes added.

## How do I test this PR?

TBD
 
## How do we deploy this PR?

TBD, there will be deployment details though. This adds new changes to the story document that need to migrate over.

nick-funk committed:
Initial implementation of story tree generation

nick-funk committed:
Implement story tree updating and generation

nick-funk committed:
Remove debugging console.log's

nick-funk committed:
Remove double await from auto-correct

nick-funk committed:
Create nextUnseenComment query

nick-funk committed:
Enforce Z_KEY for nextUnseenComment endpoint

nick-funk committed:
Limit generating story tree's to administrators

nick-funk committed:
Remove retrieval of story tree when unnecessary

This saves on bandwidth and serialization of the document since the
tree can be quite large.

nick-funk committed:
Make nextUnseenComment payload id's optional

They can come back as null, so this makes sense.

nick-funk committed:
Fix typo in story find query

nick-funk committed:
Create mutation to regenerate all story trees

nick-funk committed:
Properly resolve the rootCommentID for root comments

kabeaty commented on src/core/server/models/story/story.ts:
```diff
@@ -911,3 +938,313 @@ export async function markStoryAsUnarchived(
+  let current: StoryTreeComment | null | undefined = null;
+
+  while (stack.length > 0) {
+    if (orderBy === GQLCOMMENT_SORT.CREATED_AT_DESC) {
```
This should be checking for `CREATED_AT_ASC`

kabeaty commented on src/core/server/models/story/story.ts:
```diff
@@ -911,3 +938,313 @@ export async function markStoryAsUnarchived(
+    }
+
+    if (current.replies.length > 0) {
+      if (orderBy === GQLCOMMENT_SORT.CREATED_AT_DESC) {
```
This should be checking for `CREATED_AT_ASC` too

nick-funk committed:
Fix bad copied sort order cursor traversals

nick-funk committed:
Remove marking seen on pagination/loading replies

We will only mark seen using the `markCommentsAsSeen` mutation

nick-funk committed:
Make `nextUnseenComment` traverse relative to current comment id

nick-funk committed:
Add a bit more documentation to the next unseen search

kabeaty committed:
next unseen comment fetch

kabeaty committed:
add index to payload; check comment authorID against userID

kabeaty committed:
update to handle null currentCommentID

nick-funk committed:
Filter out non-visible comments from the stack during next unseen

We want to recreate a linear stack that matches the stream as closely
as possible.

nick-funk committed:
Prune the story tree for visibility before searching for next unseen

nick-funk committed:
Properly flatten the story tree and append root indices

nick-funk committed:
Flatten the story according to the stream sort order

nick-funk committed:
Account for viewNew comments stream side when calculating root indices

Also notifies the user if we looped around in the stream when finding
the next unseen and they need to load in the `View [n] new comments`

nick-funk committed:
Store the `viewNewCount` in local so keyboard shortcuts can use it

nick-funk committed:
Set viewNewCount in local to null when we switch stream tabs

nick-funk committed:
Update findNextUnseenVisibleCommentID traversal documentation

nick-funk committed:
Add documentation to new story tree mutation/query types

kabeaty commented on src/core/client/stream/tabs/Comments/Stream/Subscriptions/CommentEnteredSubscription.ts:
```diff
@@ -51,6 +52,11 @@ function updateForNewestFirst(
+
+  const local = store.get(LOCAL_ID);
+  if (local) {
+    local.setValue(linked.length, "viewNewCount");
```
This needs to be `linked.length + 1` to account for the new `commentsEdge`

nick-funk replied:
Fixed!

kabeaty commented on src/core/server/models/story/story.ts:
```diff
@@ -911,3 +939,501 @@ export async function markStoryAsUnarchived(
+) {
+  result.push({ ...comment, rootIndex });
+
+  if (orderBy === GQLCOMMENT_SORT.CREATED_AT_ASC) {
```
The order checks need to be switched here in `flattenComment` for replies to be included in the correct order. (So, `CREATED_AT_DESC` here and `CREATED_AT_ASC` below.)

nick-funk replied:
Rejiggered the ordering in the story tree's also fixed up this flattening algo to match what you should see in stream. Should be working now!

nick-funk committed:
Fix off by one error in comment entered subscription `viewNewCount`

nick-funk committed:
Fix ordering of story tree root/replies

- Story tree is stored in newest first (CREATED_AT_DESC)
  - This is the same for root level and replies
- Flatten tree will properly flatten root level comments
  according to the sort order given.
  - If you order by CREATED_AT_DESC, it will flatten from
    0 to n (forwards) in the root level tree comments.
  - If you order by CREATED_AT_ASC, it will flatten from
    n to 0 (backwards) in the root level tree comments.
- Flatten tree will properly flatten replies with oldest
first as the sort order does not affect replies, they are
always oldest first (CREATED_AT_ASC) order.

nick-funk committed:
Fix regenerateStoryTrees so it processes all stories properly

kabeaty committed:
dont mark new comments/replies as seen; fix linting errors

nick-funk committed:
Process the story tree regeneration using bulk operations

kabeaty commented on src/core/server/models/story/story.ts:
```diff
@@ -911,3 +939,531 @@ export async function markStoryAsUnarchived(
+  // order is descending.
+  let cursor = 0;
+  if (!currentCommentID) {
+    if (orderBy === GQLCOMMENT_SORT.CREATED_AT_ASC) {
```
I think the cursor should always just be set to `-1` in this case now so that we always start looking at 0 if there is no `currentCommentID`
